agent/consul/server.go

```diff
@@ -425,6 +404,29 @@ func NewServer(config *Config, flat Deps, publicGRPCServer *grpc.Server) (*Serve
 		fsm:                     newFSMFromConfig(flat.Logger, gc, config),
 	}
 
+	var recorder *middleware.RequestRecorder
+	if flat.NewRequestRecorderFunc != nil {
+		recorder = flat.NewRequestRecorderFunc(serverLogger, s.IsLeader, s.config.Datacenter)
+	} else {
+		return nil, fmt.Errorf("cannot initialize server without an RPC request recorder provider")
+	}
+	if recorder == nil {
+		return nil, fmt.Errorf("cannot initialize server without a non nil RPC request recorder")
+	}
+
+	var rpcServer, insecureRPCServer *rpc.Server
+	if flat.GetNetRPCInterceptorFunc == nil {
+		rpcServer = rpc.NewServer()
+		insecureRPCServer = rpc.NewServer()
+	} else {
+		rpcServer = rpc.NewServerWithOpts(rpc.WithServerServiceCallInterceptor(flat.GetNetRPCInterceptorFunc(recorder)))
+		insecureRPCServer = rpc.NewServerWithOpts(rpc.WithServerServiceCallInterceptor(flat.GetNetRPCInterceptorFunc(recorder)))
+	}
+
+	s.rpcRecorder = recorder
+	s.rpcServer = rpcServer
+	s.insecureRPCServer = insecureRPCServer
```

Comment:
keep me honest here please, but this should be 100% ok and no racing should occur given that we are still just setting up the server struct here.

Reply:
Seems fine to me. You could even remove the variable declaration and just assign directly to `s` too:
```
if flat.GetNetRPCInterceptorFunc == nil {
		s.rpcServer = rpc.NewServer()
		s.insecureRPCServer = rpc.NewServer()
} else {
		...
}
```

Reply:
Yes this should still be safe, but if we remove the dependency to isLeader we can revert it as it was.

Reply:
Per offline chats, we are going to keep `IsLeader` func arg as is for now